src/sentry/static/sentry/app/components/externalIssues/abstractExternalIssueForm.tsx

```diff
@@ -0,0 +1,235 @@
+import {t} from 'app/locale';
+import {IntegrationIssueConfig, IssueConfigField} from 'app/types';
+import FieldFromConfig from 'app/views/settings/components/forms/fieldFromConfig';
+import Form from 'app/views/settings/components/forms/form';
+import {FieldValue} from 'app/views/settings/components/forms/model';
+import {FormField} from 'app/views/settings/projectAlerts/issueEditor/ruleNode';
+
+export type ExternalIssueAction = 'create' | 'link';
+
+type Props = {
+  action: ExternalIssueAction;
+} & AsyncComponent['props'];
+
+type State = {
+  integrationDetails?: IntegrationIssueConfig;
+  dynamicFieldValues?: {[key: string]: FieldValue};
+} & AsyncComponent['state'];
+
+const DEBOUNCE_MS = 200;
+
+/**
+ * @abstract
+ */
+export default class AbstractExternalIssueForm<
+  P extends Props = Props,
+  S extends State = State
+> extends AsyncComponent<P, S> {
+  shouldRenderBadRequests = true;
+  getDefaultState(): State {
+    return {
+      ...super.getDefaultState(),
+      // This is derived from integrationDetails when it loads.
+      dynamicFieldValues: undefined,
+      // This is fetched by AsyncComponent.getEndpoints.
+      integrationDetails: undefined,
+    };
+  }
+
+  refetchConfig = () => {
+    const {action} = this.props;
+    const {dynamicFieldValues} = this.state;
+    const query = {action, ...dynamicFieldValues};
+    const endpoint = this.getEndPointString();
+
+    this.api.request(endpoint, {
+      method: 'GET',
+      query,
+      success: (data, _, jqXHR) => {
+        this.handleRequestSuccess({stateKey: 'integrationDetails', data, jqXHR}, true);
+      },
+      error: error => {
+        this.handleError(error, ['integrationDetails', endpoint, null, null]);
+      },
+    });
+  };
+
+  getConfigName = (): string => {
+    // Explicitly returning a non-interpolated string for clarity.
+    const {action} = this.props;
+    switch (action) {
+      case 'create':
+        return 'createIssueConfig';
+      case 'link':
+        return 'linkIssueConfig';
+      default:
+        throw new Error('illegal action');
+    }
+  };
+
+  getDynamicFields(
+    integrationDetails?: IntegrationIssueConfig
+  ): {[key: string]: string | undefined} {
+    const config: IssueConfigField[] = (integrationDetails ||
+      this.state.integrationDetails ||
+      {})[this.getConfigName()];
+
+    return Object.fromEntries(
+      config
+        .filter((field: IssueConfigField) => field.updatesForm)
+        .map((field: IssueConfigField) => [field.name, field.default])
+    );
+  }
+
+  onRequestSuccess({stateKey, data}) {
+    if (stateKey === 'integrationDetails' && !this.state.dynamicFieldValues) {
+      this.setState({
+        dynamicFieldValues: this.getDynamicFields(data),
+      });
+    }
+  }
+
+  onFieldChange = (label: string, value: FieldValue) => {
+    const {dynamicFieldValues} = this.state;
+
+    if (label in this.getDynamicFields() && dynamicFieldValues) {
```

Comment:
@mgaeta It's safer to use `hasOwnProperty` rather than `in` for checking non-static keys.
```suggestion
    let dynamicFields = this.getDynamicFields()
    if (dynamicFields.hasOwnProperty(label) && dynamicFieldValues) {
````